### PR TITLE
feat: Frame budgeting

### PR DESCRIPTION
**Pull request by dalkia**
- Introduces basic profiling system and view. It can be useful for the presentation, to show without a development build the performance increases. 

- Introduces a CAP Frame time budget provider, which makes systems returns if they try to act when a certain frame cap has been reachead (currently at 20 ms). It affects the following systems: 

1. LOADING SYSTEMS: `FinalizeGLTFContainerLoadingSystem`

2. INSTATIATION SYSTEM: `InstantiatePrimitiveRenderingSystem`, `StartMaterialsLoadingSystem`, `CreateBasicaMaterialSystem`,  `CreatePBRMaterialSystem`, `CreateGLTFFromAssetBundleSystem`

- Adds basic Profiling view (FPS and hiccups) for in-game metrics

### CODE REVIEW

File: Explorer/Assets/Scripts/ECS/Profiling/Systems/ProfilerWindowSystem.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using Cysharp.Threading.Tasks;
+using ECS.Abstract;
+using System.Threading;
+using UnityEngine;
+using Utility;
+
+namespace ECS.Profiling.Systems
+{
+    [UpdateInGroup(typeof(InitializationSystemGroup))]
+    public partial class ProfilerWindowSystem : BaseUnityLoopSystem
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Profiling/Systems/ProfilerWindowSystem.cs:13]: You call it a system but it does not behave as a system

File: Explorer/Assets/Scripts/ECS/Profiling/Systems/ProfilerWindowSystem.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        private void HideMetrics()
+        {
+            cts.SafeCancelAndDispose();
+        }
+
+        protected override void Update(float t) { }
+
+        private async UniTaskVoid UpdateView(CancellationToken ct)
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Profiling/Systems/ProfilerWindowSystem.cs:42]: It's very strange: you should have this logic in `Update` function. Also this way you're hiding the impact of the system from `Profiler`

File: Explorer/Assets/Scripts/ECS/Profiling/ProfilingProvider.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            mainThreadTimeRecorder.CurrentValue;
+
+        public double GetAverageFrameTimeValueInNS() =>
+            GetRecorderFPSAverage(mainThreadTimeRecorder);
+
+
+        public int GetHiccupCountInBuffer() =>
+            hiccupBufferCounter.HiccupsCountInBuffer;
+
+        private async UniTaskVoid CountHiccup()
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Profiling/ProfilingProvider.cs:34]: It should be handled in a system. Only systems are authorized to have loops like this

File: Explorer/Assets/Scripts/ECS/SceneLifeCycle/SceneDefinition/Systems/LoadSceneDefinitionListSystem.cs [hunk 4]
```diff
@@ -32,7 +32,7 @@ public partial class LoadSceneDefinitionListSystem : LoadSystemBase<SceneDefinit
         // There is no cache for the list but a cache per entity that is stored in ECS itself
         internal LoadSceneDefinitionListSystem(World world, IStreamableCache<SceneDefinitions, GetSceneDefinitionList> cache,
             MutexSync mutexSync)
-            : base(world, cache, mutexSync) { }
+            : base(world, cache, mutexSync, new NullBudgetProvider()) { }
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/SceneLifeCycle/SceneDefinition/Systems/LoadSceneDefinitionListSystem.cs:35]: Also if you really want to budget some specific loading systems (and their synchronous logic) it does not look sensible to include `BudgetProvider` in a base class.

File: Explorer/Assets/Scripts/Utility/CancellationTokenExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+using System;
+using System.Threading;
+
+namespace Utility
+{
+    public static class CancellationTokenExtensions
+    {
+        public static void SafeCancelAndDispose(this CancellationTokenSource cancellationToken)
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/CancellationTokenExtensions.cs:8]: You won't need it if proposed changes are applied

File: Explorer/Assets/Scripts/ECS/Unity/GLTFContainer/Asset/Systems/CreateGltfAssetFromAssetBundleSystem.cs [hunk 6]
```diff
@@ -71,11 +66,6 @@ private void ConvertFromAssetBundle(in Entity entity, ref GetGltfContainerAssetI
                 return;
             }
 
-            if (!instantiationThrottler.Acquire(assetBundleData.GameObjectNodes.Count))
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Unity/GLTFContainer/Asset/Systems/CreateGltfAssetFromAssetBundleSystem.cs:74]: Do you remember we were talking about removing the list and replacing it with a single object to exclude bias and not break each iteration by budgeting?

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]

File: Explorer/Assets/Scripts/ECS/Profiling/ProfilingProvider.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            var samplesCount = recorder.Capacity;
+            if (samplesCount == 0)
+                return 0;
+
+            double r = 0;
+            unsafe
+            {
+                var samples = stackalloc ProfilerRecorderSample[samplesCount];
+                recorder.CopyTo(samples, samplesCount);
```
[the remark this answers is not in the document]
Comment (dalkia) [reply Explorer/Assets/Scripts/ECS/Profiling/ProfilingProvider.cs:53]: Changed! I saw the example in the Unity API (https://docs.unity3d.com/ScriptReference/Unity.Profiling.ProfilerRecorder.html); I missed the GetSample method

**Review by mikhail-dcl — APPROVED** [review]
GJ

**Pull request merged**